cbrain_task/fsl_melodic/bourreau/fsl_melodic.rb

```diff
@@ -50,82 +51,448 @@ def job_walltime_estimate #:nodoc:
-    cmds << "echo \"set fmri\(multiple\) 1\" >> #{modified_design_file}\n"
+      # Converts minc files to nifti.
+      functional_file , functional_conversion_command  = converted_file_name_and_command functional_file_id
+      if functional_conversion_command.present?
+        cmds << functional_conversion_command 
+        params[:converted_files][functional_file_id] = functional_file
+      end
 
-    cmd_melodic = "feat #{modified_design_file}"
-    cmds    << "echo running #{cmd_melodic.bash_escape}"
-    cmds    << cmd_melodic
+      # Performs auto-correction based on the first functional file. 
+      unless auto_correction_done
 
-    params[:output_dir_name] = output
+        auto_correction_done = true
+        
+        # Auto-correction needs to be done on
+        # the task node (i.e. in the task script), not on the Bourreau
+        # host.  Otherwise, FSL needs to be installed on the Bourreau
+        # node, which might not be the case.
+        
+        if params[:npts_auto] == "1"
+          cmds << find_command("FSLNVOLS","fslnvols fsl5.0-fslnvols")
+          cmds << "# Auto-corrects parameter fmri(npts)\n"
+          cmds << "NPTS=`${FSLNVOLS} #{functional_file}`\n"
+          cmds << "if [ $? != 0 ]\n"
+          cmds << "then\n"
+          cmds << "  echo ERROR: cannot auto-correct number of volumes in #{functional_file} '(fslnvols failed)'.\n"
+          cmds << "  exit 1\n"
+          cmds << "fi\n"
+          cmds << "echo Auto-corrected number of volumes to ${NPTS}.\n"
+          cmds << set_design_file_option(modified_design_file_path,"npts","${NPTS}")
+        end
+        
+        
+        if params[:tr_auto] == "1"
+          cmds << find_command("FSLHD","fslhd fsl5.0-fslhd")
+          cmds << "# Auto-corrects parameter fmri(tr)\n"
+          cmds << "TR=`${FSLHD} #{functional_file} | awk '$1==\"pixdim4\" {print $2}'`\n"
+          cmds << "if [ $? != 0 ]\n"
+          cmds << "then\n"
+          cmds << "  echo ERROR: cannot auto-correct TR in #{functional_file} '(fslhd failed)'.\n"
+          cmds << "  exit 1\n"
+          cmds << "fi\n"
+          cmds << "echo Auto-corrected TR to ${TR}.\n"
+          cmds << set_design_file_option(modified_design_file_path,"tr","${TR}")
+        end
+        
+        if params[:totalvoxels_auto] == "1"
+          cmds << find_command("FSLSTATS","fslstats fsl5.0-fslstats")
+          cmds << "# Auto-corrects parameter fmri(totalVoxels)\n"
+          cmds << "NVOX=`${FSLSTATS} #{functional_file} -v | awk '{print $1}'`\n"
+          cmds << "if [ $? != 0 ]\n"
+          cmds << "then\n"
+          cmds << "  echo ERROR: cannot auto-correct number of voxels in #{functional_file} '(fslstats failed)'.\n"
+          cmds << "  exit 1\n"
+          cmds << "fi\n"
+          cmds << "echo Auto-corrected total voxels to ${NVOX}.\n"
+          cmds << set_design_file_option(modified_design_file_path,"totalVoxels","${NVOX}")
+        end
```

Comment:
Wouldn't using heredocs instead of array insertions would make the code easier to read and reason about?